### PR TITLE
Split Executor.Run to Executor.Prepare and Executor.RunPreparedContext for inference

### PR DESCRIPTION
**Pull request by Xreki**
In #9000, the basic `Executor.Run` was split into two low-level functions, `Executor.Prepare` and `Executor.RunPreparedContext`. In `Executor.Prepare`, operators are created in advance and saved in a `ExecutorPrepareContext`. We can always call `Executor.RunPreparedContext` to use the `ExecutorPrepareContext` instead to avoid frequently creating and destroying of operators if there is no change to the program.

For inference, a high level [Executor.Run](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/executor.h#L54) is defined for direct use in C++ inference codes. We should also provide a corresponding high level `Executor.RunPreparedContext` if users want to avoid the cost of frequently creating and destroying of operators.

### CODE REVIEW

File: paddle/fluid/framework/executor.cc [hunk 1]
```diff
@@ -129,13 +129,15 @@ static bool has_feed_operators(
...
 
-    // When feed operator are present, so should be feed_holder
-    auto var = block.FindVar(feed_holder_name);
-    PADDLE_ENFORCE_NOT_NULL(var, "Block should already have a '%s' variable",
-                            feed_holder_name);
-    PADDLE_ENFORCE_EQ(var->GetType(), proto::VarType::FEED_MINIBATCH,
-                      "'%s' variable should be 'FEED_MINIBATCH' type",
-                      feed_holder_name);
+    if (!feed_holder_name.empty()) {
+      // When feed operator are present, so should be feed_holder
```
Comment (luotao1) [inline paddle/fluid/framework/executor.cc]: - When feed operators are present
- so should be feed_holder缺少主语？

File: paddle/fluid/inference/tests/test_helper.h [hunk 2]
```diff
@@ -178,7 +185,14 @@ void TestInference(const std::string& dirname,
           "run_inference",
           paddle::platform::DeviceContextPool::Instance().Get(place));
 
-      executor.Run(*inference_program, scope, feed_targets, fetch_targets);
+      if (PrepareContext) {
+        // Note: if you changed the inference_program, you need to call
```
Comment (luotao1) [inline paddle/fluid/inference/tests/test_helper.h]: changed->change

**Review by luotao1 — COMMENTED** [review]
可以在第一个comment里面补充一些这个PR的背景么，即为什么要拆分成两个？

File: paddle/fluid/framework/executor.cc (hunk 1 above)
Comment (Xreki) [reply paddle/fluid/framework/executor.cc]: 这似乎是一种英语句法：https://zhidao.baidu.com/question/6942604.html

Comment (Xreki) [reply paddle/fluid/framework/executor.cc]: 这似乎是一种英语句法：https://zhidao.baidu.com/question/6942604.html

File: paddle/fluid/inference/tests/test_helper.h (hunk 2 above)
Comment (Xreki) [reply paddle/fluid/inference/tests/test_helper.h]: Done.

**Review by luotao1 — APPROVED** [review]
LGTM

**Pull request merged**